### PR TITLE
Aww 1153 dex freeze eth/pax create/cancel order

### PR DESCRIPTION
**Pull request by leonidstr**

### CODE REVIEW

File: doc/apl-blockchain.properties [hunk 1]
```diff
@@ -332,6 +332,8 @@ apl.eth.node.url=https://ropsten.infura.io/v3/fd015ebfa03c41c1a3d4a24734da99b4
 
 apl.eth.pax.contract.address=0xd4aecf650c2c9a0d4e2bf08fd10268ca1e199fcd
 
+apl.eth.smart.contract.address=0xd811b4E6164B889Ed15196c14A051EA3a8EDF943
```
Comment (AndrewBoyarsky) [inline doc/apl-blockchain.properties]: Maybe rename property to 'apl.eth.contract.address'. Could you add some description for these properties?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexService.java [hunk 2]
```diff
@@ -31,23 +30,29 @@
...
 
     @Inject
-    public DexService(EthereumWalletService ethereumWalletService, DexOfferDao dexOfferDao, DexOfferTable dexOfferTable, TransactionProcessorImpl transactionProcessor) {
+    public DexService(EthereumWalletService ethereumWalletService, DexOfferDao dexOfferDao, DexOfferTable dexOfferTable, TransactionProcessorImpl transactionProcessor,
+                      DexOfferTransactionCreator dexOfferTransactionCreator, DexSmartContractService dexSmartContractService) {
         this.ethereumWalletService = ethereumWalletService;
         this.dexOfferDao = dexOfferDao;
         this.dexOfferTable = dexOfferTable;
         this.transactionProcessor = transactionProcessor;
+        this.dexOfferTransactionCreator = dexOfferTransactionCreator;
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexService.java]: It is safer to add null checks here for all injected parameters.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import javax.inject.Singleton;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.concurrent.ExecutionException;
+
+import static com.apollocurrency.aplwallet.apl.util.Constants.ETH_DEFAULT_ADDRESS;
+
+@Singleton
+public class DexSmartContractService {
+    private static final Logger LOG = LoggerFactory.getLogger(DexSmartContractService.class);
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java]: Can be replaced by lombok @Slf4j annotation

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    private String smartContractAddress;
+    private String paxContractAddress;
+    private KeyStoreService keyStoreService;
+    private DexEthService dexEthService;
+    private EthereumWalletService ethereumWalletService;
+
+    @Inject
+    public DexSmartContractService(Web3j web3j, PropertiesHolder propertiesHolder, KeyStoreService keyStoreService, DexEthService dexEthService,
+                                   EthereumWalletService ethereumWalletService) {
+        this.web3j = web3j;
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java]: Pls add required parameters check

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    /**
+     *  Deposit(freeze) money(eth or pax) on the contract.
+     * @param currency Eth or Pax
+     * @return String transaction hash.
+     */
+    public String deposit(String passphrase, long accountId, String fromAddress, BigInteger weiValue, Long gas, DexCurrencies currency) throws ExecutionException, AplException.ExecutiveProcessException {
+        WalletKeysInfo keyStore = keyStoreService.getWalletKeysInfo(passphrase, accountId);
+        EthWalletKey ethWalletKey = keyStore.getEthWalletForAddress(fromAddress);
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java]: NPE is possible
keyStoreService.getWalletKeysInfo(passphrase, accountId)
can return NULL

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+
+    /**
+     *  Withdraw money(eth or pax) from the contract.
+     * @param currency Eth or Pax
+     * @return String transaction hash.
+     */
+    public String withdraw(String passphrase, long accountId, String fromAddress,  BigInteger weiValue, Long gas, DexCurrencies currency) throws AplException.ExecutiveProcessException {
+        WalletKeysInfo keyStore = keyStoreService.getWalletKeysInfo(passphrase, accountId);
+        EthWalletKey ethWalletKey = keyStore.getEthWalletForAddress(fromAddress);
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java]: Seems NPE is possible

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+        return withdraw(ethWalletKey.getCredentials(), weiValue, gasPrice, currency.isEth() ? null : paxContractAddress);
+    }
+
+    /**
+     * Get information how much money frozen a particular address.
+     * @return Amount of the froze money on the contract for the address.
+     */
+    public BigDecimal getUserInfoEth(String address, DexCurrencies currency){
+        if(!currency.isEthOrPax()){
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java]: NPE can happen without incoming param check.

Comment (leonidstr) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexSmartContractService.java]: It's impossible,  currency is required parameter.  All validation on the Controller layer.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/dao/EthGasStationInfoDao.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+@Singleton
+public class EthGasStationInfoDao {
+    private static final Logger LOG = LoggerFactory.getLogger(EthGasStationInfoDao.class);
+    private static String ETH_GAS_INFO_URL = "https://ethgasstation.info/json/ethgasAPI.json";
```
[the remark this answers is not in the document]
Comment (leonidstr) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/dao/EthGasStationInfoDao.java]: I don't think we will use it from another place.

**Pull request merged**